feat(s3): add S3Storage adapter

karfau opened the pull request:
- currently only works for deno, but dependencies ready for node
- is based on the jsr package [`@bradenmacdonald/s3-lite-client@0.7.5`](https://github.com/bradenmacdonald/s3-lite-client), which is not available on npm
  - added JSR registry to `.npmrc`
  - using `workspace:*` as version specifier to stop pnpm complaining about private workspace packages after adding the JSR registry

https://pnpm.io/workspaces#referencing-workspace-packages-through-aliases
https://jsr.io/docs/npm-compatibility

karfau committed:
feat(s3): add S3Storage adapter

- currently only works for deno
- is based on the jsr package `@bradenmacdonald/s3-lite-client@0.7.4`, which is not available on npm

Satont commented:
You can unignore `.npmrc` and make necessary changes.

But i'm not sure what to do with `storage/utils` since it's a helper package which doesn't implies publishing it somewhere. Does it have solution? I'm not familiar with jsr.

Satont commented on packages/s3/README.md:
````diff
@@ -0,0 +1,113 @@
+   ```ts
+   const getSessionKey = (ctx: MyContext) =>
+     // it could be user based
+     `/chat/${ctx.from?.id ?? 0}/sesssion.json`;
````
Typo i guess?

karfau committed:
fix: sync libs/utils/src/deps.*.ts

Satont commented on packages/s3/test/adapter.test.ts:
```diff
@@ -0,0 +1,135 @@
+} from '../adapter.ts';
+import { lazySession } from 'https://lib.deno.dev/x/grammy@1.x/mod.ts';
+import { createBot, createMessage } from '../../../libs/utils/src/mod.ts';
+import { assertSpyCall, spy } from 'jsr:@std/testing/mock';
+import { assert } from 'jsr:@std/assert';
```
Is it deno related imports?

karfau replied:
Yes, is that a problem? wo you prefer an `imports` field in `deno.json`?

Satont replied:
> Yes, is that a problem? wo you prefer an `imports` field in `deno.json`?

I guess so for std'ed deps.

karfau replied:
:heavy_check_mark:

Satont commented on packages/s3/adapter.ts:
```diff
@@ -0,0 +1,72 @@
+import { StorageAdapter } from 'https://lib.deno.dev/x/grammy@1.x/mod.ts';
+import { S3Client } from 'jsr:@bradenmacdonald/s3-lite-client@0.7.4';
+export { S3Client } from 'jsr:@bradenmacdonald/s3-lite-client@0.7.4';
```
Import it from deps file, so then when we'll nodefy this adapter dependencies will be used from deps.node.ts.

karfau replied:
makes sense

karfau replied:
:heavy_check_mark:

Satont commented on packages/s3/adapter.ts:
```diff
@@ -0,0 +1,72 @@
+S3Client,
+'exists' | 'deleteObject' | 'getObject' | 'host' | 'region' | 'putObject'
+>;
+const isS3StorageClient = (
+  maybeClient: S3StorageClient | any,
+): maybeClient is S3StorageClient =>
+  ['exists', 'deleteObject', 'getObject', 'putObject'].every((required) =>
+    typeof maybeClient[required] === 'function'
+  );
```
Please use code block.

Satont replied:
> I'm not sure I understand this comment, do you want me to add braces around the function body and add a return statement? If not, can you provide an example of what you mean?

Yep, thats what i meant. I'm personaly not an fan of such expressions.

karfau replied:
I guess you can put this into prettier(?) or eslint config, so you don't have to remind people ...

karfau replied:
✔️

Satont commented on packages/s3/adapter.ts:
```diff
@@ -0,0 +1,72 @@
+  /**
+   * @see https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html
+   */
+  readonly isSafe = (key: string): boolean => /^[-0-9a-zA-Z!_.()]+$/.test(key);
+
+  readonly delete = (key: string): Promise<void> =>
+    this.client.deleteObject(key);
+  readonly has = (key: string): Promise<boolean> => this.client.exists(key);
+  readonly read = async (key: string): Promise<T | undefined> => {
+    try {
+      const res = await this.client.getObject(key);
+      const data = await res.json() as T;
+      return this.validateSession(data) ? data : undefined;
+    } catch {
+      return undefined;
+    }
+  };
+
+  readonly write = async (key: string, value: T): Promise<void> => {
+    // the client has a mismatching return type
+    // to make type checks happy we await it, and intentionally do not return it
+    await this.client.putObject(key, JSON.stringify(value));
+  };
```
Why is it arrow functions?

karfau replied:
So that you can pass them around and without loosing the `this` context.
But if you prefer I can make them class methods.

karfau replied:
changed it to class methods

karfau replied:
somehow I missed some of them earlier. Now all of them are async class methods that await their promises.

Satont commented:
> I will see what I can find regarding the issues with jsr and pnpm workspaces. But can you tell me which version of pnpm to use? The current lockfile has version 6. when I user the latest pnpm it upgrades it to version 9 and has tons of changes...
> 
> Would it make sense to set the package manager version in package.json, so it can simply be installed using `corepack enable`?

yea i guess it need to be updated to pnpm 9 version, i'll add changes to main branch. Thanks for pointing it.

karfau committed:
docs: resolve issues from review

karfau committed:
chore: move dependencies to import map/deps.deno.ts

karfau commented on packages/s3/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2022-2024 Satont
```
:heavy_check_mark:

karfau committed:
chore: enable usage of jsr packages

- add JSR registry to `.npmrc`
- using `workspace:*` as version specifier to stop pnpm complaining about private workspace packages

https://pnpm.io/workspaces#referencing-workspace-packages-through-aliases
https://jsr.io/docs/npm-compatibility

karfau committed:
refactor: use class methods and code blocks

and apply deno fmt

karfau committed:
refactor: use class methods everywhere and await all promises

karfau committed:
docs: avoid lambda in docs

karfau committed:
docs: avoid lambda in docs

karfau committed:
style: apply deno fmt

karfau commented on packages/s3/adapter.ts:
```diff
@@ -0,0 +1,72 @@
+  );
+
+export const isObjectSession = (maybeSession: any): maybeSession is object =>
+  !!maybeSession && typeof maybeSession === 'object';
```
:heavy_check_mark:

karfau committed:
docs: add s3 to commit scopes

and sort them alphabetically

karfau committed:
style: apply prettier to s3 package

karfau commented:
I did see that the check failed, might only be able to take care of it next week

karfau committed:
test: fix type issue

karfau committed:
refactor: bump @bradenmacdonald/s3-lite-client

and remove vendored type S3StorageClient

karfau commented:
@Satont I was able to resolve all todos in the code and I think I also managed to cater for all your comments. 

If there is anything else required for the PR to land, let me know.

karfau commented:
Is there an ETA for landing/reviewing this PR?

Satont commented:
Hello. Is this still alive? I can write adapter myself, if not.

Satont commented:
I'll implement it myself.
